src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs

```diff
@@ -117,10 +117,31 @@ private Task<bool> NominateProjectAsync(string projectUniqueName, IVsProjectRest
                     $"The nominate API is called for '{projectUniqueName}'.");
 
                 var projectNames = ProjectNames.FromFullProjectPath(projectUniqueName);
+                DependencyGraphSpec dgSpec;
+                IReadOnlyList<IAssetsLogMessage> nominationErrors = null;
+                try
+                {
+                    dgSpec = ToDependencyGraphSpec(projectNames, projectRestoreInfo, projectRestoreInfo2);
+                }
+                catch (Exception e)
```

Comment:
NU1010 error will be a wrapper error for anything falling here (invalid TFM, etc).  Maybe it would be worth to have a NuGet exception that will allow different project data inspectors/evaluators to send their own NU**** messages ?

Reply:
I want nomination to be resilient to unknown and future errors. There are multiple improvements we can make to this. For example, more specific NU codes for specific things like invalid project TFM or invalid version on a package ref. Another example is right now the error is thrown as soon as the first error is found. So if there are multiple problems, only the first is reported. Ideally that should be improved as well, but it can be done in another PR.

Reply:
I think adding more granular errors will not have a great ROI at this point in time.
These bad input situations are very rare. We should only invest once we know the benefit is there. 

I imagine it'd be super painful to rework it.

Reply:
For CPVM, we need to fail restore with specific NUxxx error code. One way that I was thinking to achieve this is to do the validation and throw the exception with the NUxxx error code. This is only one example, but probably there are other cases with the same pattern.

Reply:
> For CPVM, we need to fail restore with specific NUxxx error code.

Why? It's just like any input error.

Reply:
@cristinamanum I'm resolving this comment, but I'm a bit concerned why you think CPVM should duplicate all existing errors and give them new codes. Perhaps we should talk about that.

Reply:
I was not saying CPVM should duplicate all the existing errors and give them new error codes. I was referring to the fact that this CPVM error will be added with NU error code based on a  suggestion [here](https://github.com/NuGet/NuGet.Client/pull/3198#discussion_r369864252 ) to have a NU** code. 

If we decide that any input error we treat it as same NU error code is perfectly fine. However if we like to have the flexibility to differentiate the user inputs in different NU error codes the current design approach will not help. 

The current PR can go ahead and be merged as the above topic will probably need more discussions,

Reply:
Ok, there was a misunderstanding. Since CPVM downgrade errors are using a different NU code to non-CPVM downgrade warnings (😢 but at least it's more understandable because of errors vs warnings), I mis-interpreted "*For CPVM, we need to fail restore with specific NUxxx error code*" to mean that you wanted all NU codes duplicated to be different in cpvm and non-cpvm projects.

Instead, I now understand you just mean that when errors occur, you want the specific error to be reported, not NU1010 or NU1105 for all different reasons. As I mentioned in my first reply to your comment on this line of code, this is an improvement for another PR. I agree with Nikolche that there is insufficient benefit at this time.

Also, in case it's not clear, this error handling is only run when converting `IVsProjectRestoreInfo` from project system nomination into a DGSpec. Errors that get reported during `RestoreCommand` will retain their NU code, and it will also work from both command line restores and VS restores. This code that was commented on is only handling VS nominations.